[GTK] PlatformDisplayWayland is no longer used for compositing

carlosgcampos opened the pull request:
#### 50f4156bfc49b5dd29108de5d69bfc79a5a585aa
<pre>
[GTK] PlatformDisplayWayland is no longer used for compositing
<a href="https://bugs.webkit.org/show_bug.cgi?id=255726">https://bugs.webkit.org/show_bug.cgi?id=255726</a>

Reviewed by Adrian Perez de Castro.

We use either WPE renderer or DMABuf (GBM or Surfaceless) so we can
remove all the code used for compositing with PlatformDisplayWayland.

* Source/WebCore/SourcesGTK.txt:
* Source/WebCore/platform/graphics/egl/GLContext.cpp:
(WebCore::GLContext::createWindowContext):
(WebCore::GLContext::create):
(WebCore::GLContext::createOffscreen):
(WebCore::GLContext::createSharing):
(WebCore::GLContext::~GLContext):
* Source/WebCore/platform/graphics/egl/GLContextWayland.cpp: Removed.
* Source/WebCore/platform/graphics/wayland/PlatformDisplayWayland.cpp:
(WebCore::PlatformDisplayWayland::create):
(WebCore::PlatformDisplayWayland::~PlatformDisplayWayland):
(WebCore::PlatformDisplayWayland::initializeEGLDisplay):
(WebCore::PlatformDisplayWayland::initialize): Deleted.
(WebCore::PlatformDisplayWayland::registryGlobal): Deleted.
(WebCore::PlatformDisplayWayland::createSurface const): Deleted.
* Source/WebCore/platform/graphics/wayland/PlatformDisplayWayland.h:

Canonical link: <a href="https://commits.webkit.org/263231@main">https://commits.webkit.org/263231@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2ce6f4d814dbcefb4781ba3475b442fc67eadc4b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/4004 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4100 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4212 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5446 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/4253 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3981 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4191 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4075 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/4193 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/4054 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/4241 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3610 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/5372 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/1743 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3586 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/5696 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/3567 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3646 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/5091 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/4055 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/3262 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3570 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3586 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3622 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/460 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3850 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebCore/platform/graphics/egl/GLContext.cpp:
```diff
@@ -260,21 +260,24 @@ std::unique_ptr<GLContext> GLContext::createWindowContext(GLNativeWindowType win
     case PlatformDisplay::Type::Surfaceless:
         RELEASE_ASSERT_NOT_REACHED();
+        break;
     }
```
How about:

```suggestion
    default:
        RELEASE_ASSERT_NOT_REACHED();
        break;
    }
```

Or alternatively, if you want a compiler warning whenever a new PlatformDisplay::Type is added:

```suggestion
    }

    RELEASE_ASSERT(surface);
```

carlosgcampos replied:
Yes, I don't wat to use default: and I don't understand the second suggestion

aperezdc replied:
I would rather do this:

```suggestion
#if PLATFORM(WAYLAND)
    case PlatformDisplay::Type::Wayland:
#endif
#if USE(GBM)
    case PlatformDisplay::Type::GBM:
#endif
    case PlatformDisplay::Type::Surfaceless:
        RELEASE_ASSERT_NOT_REACHED();
        break;
    }
```

This is borth avoiding the duplicated code, and without using a `default:` case.

carlosgcampos replied:
That's what I had initially, but I wasn't sure it was clear enough with all the ifdefs

mcatanzaro commented on Source/WebCore/platform/graphics/wayland/PlatformDisplayWayland.cpp:
```diff
@@ -47,34 +47,19 @@
-    auto platformDisplay = std::unique_ptr<PlatformDisplayWayland>(new PlatformDisplayWayland(display));
-    platformDisplay->initialize();
-    return platformDisplay;
+    return std::unique_ptr<PlatformDisplayWayland>(new PlatformDisplayWayland(display));
```
```suggestion
    return std::make_unique<PlatformDisplayWayland>(display);
```

carlosgcampos replied:
make_unique doesn't work because the constructor is private

mcatanzaro commented on Source/WebCore/platform/graphics/wayland/PlatformDisplayWayland.cpp:
```diff
@@ -47,34 +47,19 @@
-    auto platformDisplay = std::unique_ptr<PlatformDisplayWayland>(new PlatformDisplayWayland(display));
-    platformDisplay->initialize();
-    return platformDisplay;
+    return std::unique_ptr<PlatformDisplayWayland>(new PlatformDisplayWayland(display));
```
```suggestion
    return std::make_unique<PlatformDisplayWayland>(display);
```

carlosgcampos replied:
Ditto.

mcatanzaro reviewed (commented):
Adrian would be a better reviewer for this than me, but here are a couple comments anyway.

aperezdc commented on Source/WebCore/platform/graphics/egl/GLContext.cpp:
```diff
@@ -362,38 +368,53 @@ std::unique_ptr<GLContext> GLContext::create(GLNativeWindowType window, Platform
-        return context;
+    auto context = createWindowContext(window, platformDisplay, eglSharingContext);
+    if (!context)
+        WTFLogAlways("Could not create EGL context.");
```
What's the reason to remove the last error string here? I suppose this is because `createWindowContext()` already logs the error itself.

carlosgcampos replied:
It's not removed, but moved, that error message happened when display is surfaceless, that's now handled in `GLContext::createOffscreen`.

aperezdc reviewed (approved):
LGTM, with a couple of minor comments.

carlosgcampos committed:
[GTK] PlatformDisplayWayland is no longer used for compositing
https://bugs.webkit.org/show_bug.cgi?id=255726

Reviewed by Adrian Perez de Castro.

We use either WPE renderer or DMABuf (GBM or Surfaceless) so we can
remove all the code used for compositing with PlatformDisplayWayland.

* Source/WebCore/SourcesGTK.txt:
* Source/WebCore/platform/graphics/egl/GLContext.cpp:
(WebCore::GLContext::createWindowContext):
(WebCore::GLContext::create):
(WebCore::GLContext::createOffscreen):
(WebCore::GLContext::createSharing):
(WebCore::GLContext::~GLContext):
* Source/WebCore/platform/graphics/egl/GLContextWayland.cpp: Removed.
* Source/WebCore/platform/graphics/wayland/PlatformDisplayWayland.cpp:
(WebCore::PlatformDisplayWayland::create):
(WebCore::PlatformDisplayWayland::~PlatformDisplayWayland):
(WebCore::PlatformDisplayWayland::initializeEGLDisplay):
(WebCore::PlatformDisplayWayland::initialize): Deleted.
(WebCore::PlatformDisplayWayland::registryGlobal): Deleted.
(WebCore::PlatformDisplayWayland::createSurface const): Deleted.
* Source/WebCore/platform/graphics/wayland/PlatformDisplayWayland.h:

Canonical link: https://commits.webkit.org/263231@main
